Fix #6618

mjfwebb opened the pull request:
# Description

Fixes the case where Connection queries with `@limit` in the type would generate unnecessary cypher with extra node/edge projection.

## Complexity

Low

# Issue

Closes #6618

mjfwebb committed:
test: add tests for 6618

mjfwebb committed:
refactor: remove unused method

mjfwebb committed:
feat: only project node/edge fields explicitly or on pageInfo selection

mjfwebb committed:
test: update existing test for minimal edge projection

mjfwebb committed:
docs: add changeset

neo4j-team-graphql commented:
# Performance Report

| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
| 🟩 query.QueryWithNestedIn | 12278 | 14480 | 57 | 81 | 1226 |


🟥 - Performance worsened (dbHits)
🟩 - Performance improved (dbHits)
🟦 - New test
        

<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregateWithMultipleFields | 7936 | 7936 | 60 | 68 | 1134 |
|  aggregations.NestedAggregation | 15407 | 15407 | 77 | 95 | 2174 |
|  aggregations.AggregationWithWhere | 10833 | 10833 | 51 | 66 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 20097917 | 20097917 | 1905 | 2026 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 20097917 | 20097917 | 1949 | 1960 | 2008534 |
|  aggregations.InterfacesAggregations | 8324 | 8324 | 102 | 127 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 13526 | 13526 | 122 | 157 | 2080 |
|  connections.Connection | 12951 | 12951 | 62 | 77 | 2174 |
|  connections.NestedConnection | 37705 | 37705 | 129 | 145 | 4516 |
|  connections.totalCount | 1 | 1 | 18 | 22 | 1 |
|  cypher-directive-mutation.TopLevelMutationDirective | 1135 | 1135 | 28 | 35 | 1134 |
|  cypher-directive.TopLevelSortWithCypher | 12961 | 12961 | 48 | 57 | 2174 |
|  cypher-directive.TopLevelConnectionSortWithCypher | 12961 | 12961 | 65 | 82 | 2174 |
|  cypher-directive.TopLevelSortWithCypherWithNested | 13096 | 13096 | 61 | 76 | 2174 |
|  cypher-directive.TopLevelConnectionSortWithCypherWithNested | 13096 | 13096 | 93 | 121 | 2174 |
|  cypher-directive.TopLevelSortWithExpensiveCypher | 13650 | 13725 | 116 | 143 | 2174 |
|  cypher-directive.TopLevelConnectionSortWithExpensiveCypher | 13266 | 13266 | 106 | 120 | 2174 |
|  fulltext.Fulltext | 80 | 80 | 50 | 54 | 16 |
|  fulltext.FulltextWithNestedQuery | 516 | 516 | 72 | 75 | 84 |
|  1262.NestedConnectionWhere | 8703 | 8703 | 53 | 68 | 2174 |
|  187.QueryWhere | 8564 | 8564 | 38 | 47 | 2154 |
|  2871.NestedRelationshipFilter | 19632 | 19632 | 59 | 69 | 4395 |
|  batch-create.BatchCreate | 3600 | 3600 | 90 | 102 | 600 |
|  connect.createAndConnect | 4410 | 4410 | 54 | 58 | 1134 |
|  create.SimpleMutation | 6 | 6 | 35 | 33 | 1 |
|  delete.SimpleDelete | 19401 | 19401 | 629 | 638 | 1040 |
|  delete.NestedDeleteInUpdate | 14688 | 14688 | 114 | 167 | 1938 |
|  update.NestedUpdate | 10129 | 10129 | 61 | 81 | 2002 |
|  query.SimpleQuery | 3121 | 3121 | 24 | 27 | 1040 |
|  query.SimpleQueryWithRelationship | 15031 | 15031 | 46 | 56 | 2174 |
|  query.SimpleQueryWithNestedWhere | 8713 | 8713 | 51 | 74 | 2154 |
|  query.Nested | 3988 | 3988 | 76 | 88 | 1040 |
|  query.OrFilterOnRelationships | 34756 | 36724 | 176 | 213 | 1297 |
|  query.OrFilterOnRelationshipsAndNested | 24745 | 26759 | 204 | 243 | 1297 |
| 🟩 query.QueryWithNestedIn | 12278 | 14480 | 57 | 81 | 1226 |
|  query.DeeplyNestedConnectionWhere | 8702 | 8702 | 74 | 97 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 17357 | 17357 | 132 | 155 | 1552 |
|  query.NestedWithRelationshipSingleFilters | 3808 | 3808 | 151 | 210 | 1134 |
|  sorting.SortMultipleTypes | 3436 | 3436 | 79 | 91 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 13321 | 13321 | 96 | 118 | 2174 |
|  sorting.SortOnNestedFields | 12951 | 12951 | 45 | 64 | 2174 |
|  sorting.SortDeeplyNestedFields | 39785 | 39785 | 87 | 122 | 4516 |
|  sorting.ConnectionWithSort | 3271 | 3271 | 70 | 86 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 60 | 63 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 55 | 65 | 35 |
|  unions.NestedUnion | 309975 | 309975 | 267 | 275 | 33033 |
|  unions.NestedUnionWithMissingFields | 283949 | 283949 | 255 | 280 | 33033 |

</details>
        
Old Schema Generation: 28.157s
Schema Generation: 28.645s
Old Subgraph Schema Generation: 30.587s
Subgraph Schema Generation: 32.049s

<!-- Sticky Pull Request Commentperformance -->

mjfwebb committed:
fix: typo

mjfwebb committed:
test: update snapshot

While the generated Cypher has grown, it seems to be more inline with
what the original test intended as it now also returns edges

mjfwebb committed:
test: add tck tests for aliasing cursors

mjfwebb committed:
feat: handle edge-case where only cursor is selected from edge

luffy1727 commented on packages/graphql/src/translate/queryAST/factory/Operations/ConnectionFactory.ts:
```diff
@@ -465,6 +465,15 @@ export class ConnectionFactory {
+        const selectedEdgeCursor = findFieldsByNameInFieldsByTypeNameField(resolveTreeEdgeFields, "cursor").length > 0;
+
+        // If the client requested pageInfo or edge cursor, the operation needs to project edges so page cursors can be computed. Otherwise, avoid projecting edges.
+        if (pageInfo || selectedEdgeCursor) {
```
wouldn't it be better to be explicit and check `pageInfo !== null || selectedEdgeCursor !== null`

mjfwebb replied:
In this particular case, neither of these values can be `null`:

`const pageInfo: ResolveTree | undefined`

`const selectedEdgeCursor: boolean`

We also tend not to do explicit type checks when we can rely on boolean truthiness/falsiness.

angrykoala replied:
For more context, by convention we do not use null in this project (except some cases where we need it to interface with GraphQL or other tools), we only use undefined to try and keep our sanity

angrykoala commented on packages/graphql/tests/tck/connections/projections/projections.test.ts:
```diff
@@ -159,10 +159,15 @@ describe("Relay Cursor Connection projections", () => {
-                WITH count(this1) AS totalCount
-                RETURN { totalCount: totalCount } AS var2
+                WITH collect({ node: this1, relationship: this0 }) AS edges, count(this1) AS totalCount
+                CALL (edges) {
+                    UNWIND edges AS edge
+                    WITH edge.node AS this1, edge.relationship AS this0
+                    RETURN collect({ node: { __id: id(this1), __resolveType: \\"Actor\\" } }) AS var2
+                }
+                RETURN { edges: var2, totalCount: totalCount } AS var3
             }
-            RETURN this { .title, actorsConnection: var2 } AS this"
+            RETURN this { .title, actorsConnection: var3 } AS this"
```
I'm not sure why this has changed, as it does seem to be the opposite of what the issue is about. Was this another separate bug?

mjfwebb replied:
Will also add an integration test and changelog

angrykoala commented on packages/graphql/tests/tck/connections/alias.test.ts:
```diff
@@ -141,4 +141,136 @@ describe("Connections Alias", () => {
         `);
     });
+
+    test("should alias startCursor", async () => {
```
Should these be tck tests?

All of these aliases are done in JS, which means that the cypher itself is unchanged. 

These seem to be present in the file `alias.int.test.ts`, any reason for adding them here?

mjfwebb replied:
Will move these to integration tests

angrykoala commented on packages/graphql/tests/tck/issues/6618.test.ts:
```diff
@@ -0,0 +1,77 @@
+        `;
+
+        const result = await translateQuery(neoSchema, query);
+
```
This looks good, but I'm wondering if it is worth making a more explicit check, as this issue is specifically about the generated Cypher. Maybe checking that the cypher does not contain "LIMIT"?

This is not standard practice in our TCK tests, so I'm not too opinionated in adding it, but I feel this and the integration tests don't really validate the issue is fixed :/

mjfwebb replied:
Will add an assert for the explicit lack of the collect

angrykoala commented on packages/graphql/src/translate/queryAST/factory/Operations/ConnectionFactory.ts:
```diff
@@ -465,6 +465,15 @@ export class ConnectionFactory {
 
         operation.setHasTotalCount(Boolean(totalCount || pageInfo));
 
+        // TODO: This should probably be set as an edgeField instead
```
If I'm not mistaken, the cursor is calculated in JS, so it is probably better to keep it outside of the queryAST

mjfwebb replied:
Agreed

angrykoala commented on .changeset/five-carpets-count.md:
```diff
@@ -0,0 +1,5 @@
+"@neo4j/graphql": patch
+---
+
+Fixes the case where Connection queries with @limit in the type would generate unnecessary cypher with extra node/edge projection.
```
```suggestion
Fixes the case where Connection queries with `@limit` in the type would generate unnecessary cypher with extra node/edge projection.
```

We want to avoid pinging this GitHub user :smile:

mjfwebb committed:
Update .changeset/five-carpets-count.md

Co-authored-by: angrykoala <angrykoala@outlook.es>

mjfwebb committed:
docs: remove TODO

mjfwebb committed:
Merge branch 'dev' into fix/6618-oom

mjfwebb committed:
test: move from tck to int for new test

mjfwebb committed:
docs: add changeset

mjfwebb committed:
test: add integration test to match projection test

mjfwebb committed:
test: add assertion for collect not to be included in tck

mjfwebb committed:
Merge branch 'dev' into fix/6618-oom
